Framework/Sources/FloatingPanel.swift

```diff
@@ -477,24 +520,98 @@ class FloatingPanel: NSObject, UIGestureRecognizerDelegate {
+        // starting point.
+        // There are two cases here since the user could drag the panel e.g. from .tip over .half to .full but change
+        // directions midway between .half and .full and drags the panel back downards.
+        if (animator.fractionComplete == 1) || (animator.fractionComplete == 0 && self.panGestureContext?.nextState != self.state) {
+            self.state = (animator.fractionComplete == 0) ? self.state : self.panGestureContext?.nextState ?? self.state
+            let newTargetState = self.nextPosition(with: velocity)
+            let distance = self.distance(to: newTargetState)
+
+            if newTargetState != self.panGestureContext?.nextState && newTargetState != self.state {
+                self.startNewAnimator(for: distance, with: velocity, movingTo: newTargetState)
+            }
+
+            // Since the drag is already in progress we need to 'save' the already travelled distance on the screen in
+            // order to calculate the correct translation when calcualting the fraction (-> progress) of the animator.
+            // Otherwise the panel 'jumps' from its current position to the finger's position on the screen.
+            self.panGestureContext?.alreadyMovedTranslationOfPan = translation
+        }
+
+        self.preserveContentVCLayoutIfNeeded()
+        // Determine whether or not the panel has actually moved.
+        if animator.fractionComplete != previousFractionComplete {
+            self.viewcontroller!.delegate?.floatingPanelDidMove(self.viewcontroller!)
+        }
+    }
+
+    private func startNewAnimator(for distance: CGFloat, with velocity: CGPoint, movingTo targetState: FloatingPanelPosition) {
+        guard var panGestureContext = self.panGestureContext, self.state != targetState else {
+            return
+        }
+
+        let velocityVector = (distance != CGFloat(0)) ? CGVector(dx: 0, dy: min(abs(velocity.y)/distance, 30.0)) : .zero
+
+        let animator = behavior.interactionAnimator(self.viewcontroller!, between: self.state, and: targetState, with: velocityVector)
+        animator.addAnimations { [weak self] in
+            guard let `self` = self else { return }
+            // Set layout (-> therefor the top constraint of the FloatingPanel) to the next upcoming state in order to
+            // be able to move the panel during a drag
+            self.updateLayout(to: targetState)
+        }
+        animator.addCompletion { [weak self] pos in
+            guard let `self` = self else { return }
+            if pos == .end, let animator = self.animator, !animator.isReversed {
+                self.state = panGestureContext.nextState
+            }
+            // Set layout again to ensure that the panel's top constraint corresponds with the desired state.
+            self.updateLayout(to: self.state)
+            self.animationDidComplete(at: self.state)
+        }
+
+        self.animator = animator
+        panGestureContext.nextState = targetState
```

Comment:
I don't believe this currently works. The thing with `struct`s is that they use _copy on write semantics_ which means that a new instance is create when a property is changed. Hence you need to update `self.panGestureContext` at the end of this method to persist these changes.

Reply:
You're right. Having that said, after further implementing your proposal with the `struct`, we have to add quite a lot boilerplate code since in every method we update one of the struct's properties we first need to check if `self.panGestureContext` is set (e.g. using `guard let`) and then need to update `self.panGestureContext`. 
I agree that this change as itself adds value but I'm not sure if it makes this particular class more readable, to be honest.